server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRecipientRewriteTable.java

```diff
@@ -177,9 +178,10 @@ public String asKey() {
 
     @Override
     protected String mapAddressInternal(String user, Domain domain) throws RecipientRewriteTableException {
-        Mappings mappings = retrieveMappings(user, domain)
-            .orElseGet(() -> retrieveMappings(WILDCARD, domain)
-                .orElseGet(MappingsImpl::empty));
+        Mappings mappings = OptionalUtils.or(
+            retrieveMappings(user, domain),
+            retrieveMappings(WILDCARD, domain))
```

Comment:
you resolve all values before calling `or()`, it means twice the number of queries when first one would resolve to something.

Reply:
There is a supplier version of it, you're right.